### ISSUE TITLE
Querying from cache existing items

### ISSUE THREAD

**Issue by SanjiKir** [zino-app/graphql-flutter#626]
**Describe the bug**
Current cache has 2 types of keys. 

1. Default `typenameDataIdFromObject` implementation
2. Keys created by async call using `options.toKey()`.

Let's say I have fetched a list of items e.g.

```
Query Items {
    items {
         __typename,
         id,
     }
}
```

This will be saved in the cache under 2 keys as a result of `toKey` and as `Item/someId` considering default typenameDataIdFromObject implementation. 

Then if I'd like to access the item from the cache using query by id e.g.

```
Query {
    item(\$id: String!) {
        __typename
        id
    }
}
```

I will not get it! Because in current query implementation we have 

```
    final String cacheKey = options.toKey();
    if (shouldRespondEagerlyFromCache(options.fetchPolicy) &&
          !queryResult.optimistic) {
        final dynamic data = cache.read(cacheKey);
    ....
```

However, cache.read(typenameDataIdFromObject({ 'id': id, '__typename': 'Item' });
will work.

I suggest as a fix add `cache.read(typenameDataIdFromObjec(options))` in case first read from cache was empty so we will check for the same query and existing items. Then developer can decide how does he build his cache keys and just send it with options e.g. `id` and `__typename`.

I can create a PR if maintainers support my idea.

**Comment by SanjiKir** [comment]
Based on apollo https://www.apollographql.com/docs/react/caching/cache-interaction/ and considering I missed that new query can request more data than there currently is in the cache, there can be implemented something like `readQuery`, which should match requested properties with existing on the cache and similarly to apollo throw error if data requested doesn't match. What do you guys think?

**Comment by micimize** [comment]
This is the kind of awkwardness that should be solved by the v4 cache rework and the underlying `normalize` lib.

I'd prefer anyone with this issue try `^4.0.0-alpha.6` and come to the `#support` channel on [discord](https://discord.gg/tXTtBfC) with any issues from that

**Comment by micimize** [comment]
This is likely solved on `^4.0.0-beta`
